src/types/mod.rs

```diff
@@ -368,66 +365,85 @@ impl Type {
-        hint: &'static str,
-    ) -> Result<(), Error> {
+    pub fn bind(&self, bound: Bound, hint: &'static str) -> Result<(), Error> {
         let root = self.find_root();
         let lock = root.inner_lock();
         match lock.constraint {
             // Types with bounds are more interesting: we have to recursively unify
             // the existing bound with the new one.
             Constraint::Bound {
                 bound: ref existing_bound,
-                ref final_data,
-                ..
             } => {
+                let bind_error = || Error::Bind {
+                    existing_bound: existing_bound.shallow_clone(),
+                    new_bound: bound.shallow_clone(),
+                    hint,
+                };
+
                 match (&existing_bound, &bound) {
                     // Binding a free type to anything is a no-op
                     (_, Bound::Free(_)) => Ok(()),
                     // Free types are simply dropped and replaced by the new bound
                     (Bound::Free(_), _) => {
                         let mut lock = lock.into_mutable(); // free means non-finalized
-                        lock.constraint = Constraint::Bound {
-                            bound,
-                            final_data: bound_final_data.as_ref().map(Arc::clone),
-                        };
+                        lock.constraint = Constraint::Bound { bound };
                         Ok(())
                     }
-                    (Bound::Unit, Bound::Unit) => Ok(()),
+                    // Binding complete->complete shouldn't ever happen, but if so, we just
+                    // compare the two types and return a pass/fail
+                    (Bound::Complete(ref existing_final), Bound::Complete(ref new_final)) => {
+                        if existing_final == new_final {
+                            Ok(())
+                        } else {
+                            Err(bind_error())
+                        }
+                    }
+                    // Binding an incomplete to a complete type requires recursion.
+                    (Bound::Complete(ref complete), incomplete)
+                    | (&incomplete, Bound::Complete(ref complete)) => {
+                        match (complete.bound(), incomplete) {
+                            // A unit might match a Bound::Free(..) or a Bound::Complete(..),
+                            // and both cases were handled above. So this is an error.
+                            (CompleteBound::Unit, _) => Err(bind_error()),
+                            (
+                                CompleteBound::Product(ref comp1, ref comp2),
+                                Bound::Product(ref ty1, ref ty2),
+                            )
+                            | (
+                                CompleteBound::Sum(ref comp1, ref comp2),
+                                Bound::Product(ref ty1, ref ty2),
```

Comment:
969b1a64553b93e07c3bf9618fd01dc0e6c6ef9f: Unifying sums with products is a bug that is corrected in a later commit. Amazing that no unit tests fail.